lib/rex/ui/text/shell/history_manager.rb

```diff
@@ -91,60 +123,92 @@ def readline_available?
     defined?(::Readline)
   end
 
+  def reline_available?
+    begin
+      require 'reline'
+      defined?(::Reline)
+    rescue ::LoadError => _e
+      false
+    end
+  end
+
   def clear_readline
     return unless readline_available?
 
     ::Readline::HISTORY.length.times { ::Readline::HISTORY.pop }
   end
 
-  def load_history_file(history_file)
-    return unless readline_available?
+  def clear_reline
+    return unless reline_available?
 
-    clear_readline
-    if File.exist?(history_file)
-      File.readlines(history_file).each do |e|
-        ::Readline::HISTORY << e.chomp
-      end
+    ::Reline::HISTORY.length.times { ::Reline::HISTORY.pop }
+  end
+
+  def load_history_file(context)
+    history_file = context[:history_file]
+    history = map_library_to_history(context[:input_library])
+
+    begin
+      File.open(history_file, 'r') do |f|
+        clear_library(context[:input_library])
+        f.each do |line|
+          chomped_line = line.chomp
+          if context[:input_library] == :reline && history.last&.end_with?("\\")
+            history.last.delete_suffix!("\\")
+            history.last << "\n" << chomped_line
+          else
+            history << chomped_line
+          end
+        end
+        end
+    rescue Errno::EACCES, Errno::ENOENT => e
+      $stderr.puts("Failed to open history file: #{history_file} with error: #{e}") if debug?
     end
   end
 
-  def store_history_file(history_file)
-    return unless readline_available?
+  def store_history_file(context)
+    history_file = context[:history_file]
+    history = map_library_to_history(context[:input_library])
+
+    history_diff = history.length < MAX_HISTORY ? history.length : MAX_HISTORY
+
     cmds = []
-    history_diff = ::Readline::HISTORY.length < MAX_HISTORY ? ::Readline::HISTORY.length : MAX_HISTORY
     history_diff.times do
-      entry = ::Readline::HISTORY.pop
-      cmds.push(entry) unless entry.nil?
+      entry = history.pop
+      cmds << entry.scrub.split("\n").join("\\\n")
     end
 
-    write_history_file(history_file, cmds)
+    write_history_file(history_file, cmds.reverse)
   end
 
   def switch_context(new_context, old_context=nil)
     if old_context && old_context[:history_file]
-      store_history_file(old_context[:history_file])
+      store_history_file(old_context)
     end
 
     if new_context && new_context[:history_file]
-      load_history_file(new_context[:history_file])
+      load_history_file(new_context)
     else
       clear_readline
+      clear_reline
     end
-  rescue SignalException => e
+  rescue SignalException => _e
     clear_readline
+    clear_reline
   end
 
   def write_history_file(history_file, cmds)
     write_queue_ref = @write_queue
     remaining_work_ref = @remaining_work
+
     @write_thread ||= Rex::ThreadFactory.spawn("HistoryManagerWriter", false) do
-      while (event = write_queue_ref.pop)
+      while (write_queue_ref.size > 0 && event = write_queue_ref.pop)
```

Comment:
I'm not sure if this change is cromulent; the `write_thread` needs to block for new work to complete. Won't this terminate the thread early with this change?

Reply:
My reasoning for this change was that if there is no work, the thread shuts down. When new work/events are added to the queue, a new thread is made. The thread will continue to exist for as long as there is additional work to handle, but will exist when all events are handled and will be re-created when necessary.
Otherwise, we would need to update https://github.com/rapid7/metasploit-framework/blob/36e29535a93db750303d8b319d2b219fe6032839/lib/metasploit/framework/spec/threads/suite.rb#L31
Let me know which approach is more preferred or what this line should be changed to 👍

Reply:
I think making this change helps the tests to pass, but breaks the implementation? Potentially expose a `close` API for the tests to perform cleanup? The `close` implementation can write a nil to the queue and wait for the fork to close and set the write_thread to nil?